### PR TITLE
1Password OpVault format support

### LINKED ISSUE #1462: Add support for 1Password .opvault format
## Expected Behavior

KeePassXC should either open, or at least import, the [1Password OpVault](https://support.1password.com/opvault-overview/) database format.

While merely importing the database would help in cross-platform situations, currently the KeePassXC `Entry` and `EntryAttachments` models would need to be extended quite a bit to capture the "entry" and attachment models used by OpVault. Thus, opening OpVault for _use_ is likely a stretch goal.

## Current Behavior

There does not appear to be a way to open the `1Password.opvault` directory(database), neither as a source for importing nor as a database.

## Possible Solution

I have [a working branch](https://github.com/mdaniel/keepassxc/tree/opvault-format) that does the importing part, but it currently goes no further because that was the functionality I required.

## Context

I have 1Password on my Mac and on my Android phone (they have an iOS app too, but I am not affected by that). However, it is a major pain to access that data while on Windows or on my Linux laptop. I have been following KeePassXC for quite a while, and thus it seemed like the perfect(?) platform to teach to read OpVault, partially to improve the security community and wholly to "scratch my itch."

Since this issue/feature is unsolicited, I am also prepared for a "closed-wontfix" outcome. I didn't want to open an issue until I knew the format could, in fact, be implemented just from their specification. But now that I have an implementation that at least appears to function, I thought I'd test for interest.

## Discussion

Here is where my caveats go:
* I am not a C++ developer, so I am quite certain I have not written idiomatic C++
* The code on that branch is without question not what I would submit for inclusion. It is currently in a "one file to rule them all" state since it is basically a PoC that got out of hand :-) I also haven't fired up `clang-format`, in part because that file needs to be split into components before formatting becomes a worry
* The test coverage is lousy, where even the one test I did initially write is broken because I made the code work correctly

But if that feature is something the KeePassXC community would value, then I will do all in my power to bring the code into alignment so it can land

### PR DESCRIPTION
**Pull request by mdaniel**
## Description

The [on disk format](https://support.1password.com/opvault-overview/) used by [1Password from AgileBits](https://1password.com) is mostly JSON, but that JSON contains a boatload of base64 proprietary binary blobs that are encoded and encrypted in a non-obvious way. Thankfully, AgileBits has [published the specification](https://support.1password.com/opvault-design/) for the format, which they call `opvault`, and thus it is reasonable to teach KeePassXC to ingest that disk format.

For the time being, this code _only imports_ the disk format, since writing out any changes would be a monster amount of effort and -- certainly for my purposes -- wouldn't be an obvious win.

## Motivation and context

There is no support for reading the 1Password `.opvault` format on Linux unless one is subscribed to the `1password.com` product, which for sure isn't everyone who uses 1Password. There was only recently a Windows client added for it, and I can't speak to how well it works in order to know if this PR would benefit Windows users [beyond saving them the money for a second 1Password license, of course].

Fixes #1462

## How has this been tested?

They publish [a sample file](https://cache.agilebits.com/security-kb/freddy-2013-12-04.tar.gz) with a known password presumably expressly for testing. I have a happy-path test that gets in the ballpark of 75% code coverage, with the remaining lines concerning error conditions.

There will need to be some discussion around that :point_up: since it wasn't super clear if I should just unpack that tar into `tests/data`, or have the test try and download it, or what. So this very second, the `testopvaultformat` will fail if you clone this branch and run the tests.

And, as one might suspect, I've been using this to read both my home 1Password opvault and a separate one from work, so it is good enough for a daily driver.

There is currently some kind of misunderstanding about the relationship between `Group` and `Entry`, such that searching for an `Entry` finds it, and reports it in the correct `Group`, but clicking on the `Group` in the sidebar lists no `Entry` lines. Again, the search behavior is good enough for what I wanted it to do, but I can try to fix that if it's important

## Types of changes

- ✅ New feature (non-breaking change which adds functionality)
- Add a new import button on the welcome screen
- Add a new import menu item under `Database > Import`
- Add the supporting classes for reading the `opvault` format
- And the aforementioned test case

## Checklist:
- ✅ I have read the **CONTRIBUTING** document.
- ✅ My code follows the code style of this project. (FYI, `make format` dirties 103 files that I didn't touch)
- ✅ All new and existing tests passed.
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`.
- ✅ My change requires a change to the documentation and I have updated it accordingly. <-- this one is open for discussion; I didn't look into any documentation, but I'll be glad to do so if you feel it would be helpful
- ✅ I have added tests to cover my changes.

### CODE REVIEW

**Comment by mdaniel** [comment]
Let me know if there is any way I can help that process, including if you want me to take a swing at those code complexity issues reported by CodeFactor. CLion whined about the complexity, too, but I wasn't sure if it was being overly pedantic.

Also, as mentioned, I welcome input on dealing with the sample data for the test. The unpacked `.opvault` is 804K, and of that size, the majority is attachment payloads. Downloading and unpacking it also seems reasonable, I just don't know what the opinion is on extra-repo dependencies like that

**Comment by droidmonkey** [comment]
[test](https://trello.com/c/yBosWsfe/1-test)

File: src/format/OpData01.cpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+            qWarning("Did not consume all of opdata01!");
+        } else if (it > it_end) {
+            qWarning("opdata01 iterator greater than const_end");
+        }
+
+        const QByteArray actualHmac =
+            CryptoHash::hmac(data.mid(0, data.size() - hmacSig.size()), hmacKey, CryptoHash::Algorithm::Sha256);
+        if (actualHmac != hmacSig) {
+            RETURN_OPDATA01_ERROR_INSTANCE(QString("opdata01 failed its HMAC; wanted %1 but got %2")
+                                               .arg(QString::fromUtf8(hmacSig.toHex(':')))
```
Comment (atdt) [inline src/format/OpData01.cpp]: QByteArray::toHex with a separator was [introduced in Qt 5.9](http://doc.qt.io/qt-5/qbytearray.html#toHex-1). I believe the minimum supported version (per INSTALL.md) is Qt 5.2. There are additional calls to this function in this PR.

File: src/format/OpVaultReaderBandEntry.cpp [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+        entry->setGroup(this->m_db->rootGroup());
+    }
+
+    entry->setUpdateTimeinfo(false);
+    {
+        TimeInfo ti;
+        bool timeInfoOk = false;
+        if (bandEntry.contains("created")) {
+            const qint64 createdTime(bandEntry["created"].toInt());
+            ti.setCreationTime(QDateTime::fromSecsSinceEpoch(createdTime, Qt::UTC));
```
Comment (atdt) [inline src/format/OpVaultReaderBandEntry.cpp]: QDateTime::fromSecsSinceEpoch was [introduced in Qt 5.8](http://doc.qt.io/qt-5/qdatetime.html#fromSecsSinceEpoch); prior to that it was (I believe) QDateTime::fromTime_t. There are additional calls to this function in this PR.

File: src/format/OpData01.cpp [hunk 3]
```diff
@@ -0,0 +1,143 @@
+#include "OpData01.h"
+
+#include "crypto/CryptoHash.h"
+#include "crypto/SymmetricCipher.h"
+
+#include <QMessageLogger>
+#include <QtGui/QtGui> // for qWarning() et al
+
+#define RETURN_OPDATA01_ERROR_INSTANCE(msg)                                                                            \
```
Comment (droidmonkey) [inline src/format/OpData01.cpp]: Do not use a macro for returns. The generally accepted solution is to have a QString passed in as reference to the function with the return type of bool indicating success/failure. Also this creates a mess of dangling pointers if they aren't handled correctly.

File: src/format/OpData01.cpp [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        RETURN_OPDATA01_ERROR_INSTANCE(QString("data claims it does not start with header: %1")
+                                           .arg(QString::fromUtf8(data.mid(0, header.size()).toHex())));
+    } else {
+        it += header.size();
+    }
+
+    qlonglong qlen(0LL);
+    int len;
+    {
+        for (int i = 0; i < 8 and it != it_end; ++i, ++it) {
```
Comment (phoerious) [inline src/format/OpData01.cpp]: Don't use `and`, use `&&`, this is not Python.

File: src/format/OpData01.cpp [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    if (!data.startsWith(header)) {
+        RETURN_OPDATA01_ERROR_INSTANCE(QString("data claims it does not start with header: %1")
+                                           .arg(QString::fromUtf8(data.mid(0, header.size()).toHex())));
+    } else {
+        it += header.size();
+    }
+
+    qlonglong qlen(0LL);
+    int len;
+    {
```
Comment (phoerious) [inline src/format/OpData01.cpp]: What is this scope for? I don't see any variable declarations directly inside of it.

File: src/format/OpData01.cpp [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+        // stop kidding ourselves and having to downcast len
+        len = int(qlen);
+    }
+    // qDebug("toLongLong := %lld\n", qlen);
+
+    QByteArray iv(16, '\0');
+    for (int i = 0, iv_len = 16; i < iv_len and it != it_end; ++i, ++it) {
+        char ch = *it;
+        iv[i] = ch;
```
Comment (phoerious) [inline src/format/OpData01.cpp]: No reason to declare `ch` here.

File: src/format/OpData01.cpp [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+            qlen = (ch << (8 * i)) | qlen;
+        }
+
+        // stop kidding ourselves and having to downcast len
+        len = int(qlen);
+    }
+    // qDebug("toLongLong := %lld\n", qlen);
+
+    QByteArray iv(16, '\0');
+    for (int i = 0, iv_len = 16; i < iv_len and it != it_end; ++i, ++it) {
```
Comment (phoerious) [inline src/format/OpData01.cpp]: Use `&&`.

File: src/format/OpData01.cpp [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        RETURN_OPDATA01_ERROR_INSTANCE(QString("short IV read, wanted %1 but got %2").arg(16).arg(iv.size()));
+    }
+
+    SymmetricCipher cipher(SymmetricCipher::Aes256, SymmetricCipher::Cbc, SymmetricCipher::Decrypt);
+    if (!cipher.init(key, iv)) {
+        RETURN_OPDATA01_ERROR_INSTANCE(QString("Unable to init cipher for opdata01: %1").arg(cipher.errorString()));
+    }
+
+    QByteArray* clearText;
+    {
```
Comment (phoerious) [inline src/format/OpData01.cpp]: Okay, this scope has variable declarations, but what's it for?

File: src/format/OpData01.cpp [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        const int blockSize = cipher.blockSize();
+        // qDebug("cipher.blockSize = %d\n", blockSize);
+
+        int randomBytes = blockSize - (len % blockSize);
+        // qDebug("cipher.randomBytes1 = %d\n", randomBytes);
+        if (0 == randomBytes) {
+            // add random blk
+            randomBytes = blockSize;
+        }
+        int clear_len = len + randomBytes;
```
Comment (phoerious) [inline src/format/OpData01.cpp]: Don't use underscores in variable names.

File: src/format/OpData01.cpp [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        // qDebug("cipher.randomBytes2 = %d\tclearText.len=%d\n", randomBytes, clear_len);
+        QByteArray qbaCT(clear_len, '\0');
```
Comment (phoerious) [inline src/format/OpData01.cpp]: `qbaCT` is not a good variable name. It's not obvious what it means.

File: src/format/OpData01.cpp [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+        for (int i = 0; i < clear_len and it != it_end; ++i, ++it) {
```
Comment (phoerious) [inline src/format/OpData01.cpp]: Use `&&`.

File: src/format/OpData01.cpp [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+            qbaCT[i] = *it;
+        }
+
+        const int hmacLen = 256 / 8;
```
Comment (phoerious) [inline src/format/OpData01.cpp]: This is always 32. Any reason for not writing that directly?

File: src/format/OpData01.cpp [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+        QByteArray hmacSig; // 256 / 8, '\0');
+        for (int i = 0; i < hmacLen and it != it_end; ++i, ++it) {
```
Comment (phoerious) [inline src/format/OpData01.cpp]: Use `&&`.

File: src/format/OpData01.cpp [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+
+OpData01* OpData01::decode01(const QByteArray& data, const QByteArray& key, const QByteArray& hmacKey)
+{
+    // qDebug("OpData01.decode-ing %d bytes using a %d sized key, %d sized HMAC\n", data.size(), key.size(),
+    // hmacKey.size());
+
+    const QByteArray header("opdata01");
+
+    QByteArray::const_iterator it = data.cbegin();
+    QByteArray::const_iterator it_end = data.cend();
```
Comment (phoerious) [inline src/format/OpData01.cpp]: Don't use underscores in variable names.

File: src/format/OpData01.cpp [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+                                               .arg(QString::fromUtf8(actualHmac.toHex(':'))));
+        }
+
+        if (!cipher.processInPlace(qbaCT)) {
+            RETURN_OPDATA01_ERROR_INSTANCE(QString("Unable to process clearText in place"));
+        }
+        const QByteArray& actualClearText = qbaCT.mid(randomBytes);
+        if (actualClearText.size() != len) {
+            RETURN_OPDATA01_ERROR_INSTANCE(
+                QString("Expected %1 bytes of clear-text, found %2").arg(len).arg(actualClearText.size()));
```
Comment (phoerious) [inline src/format/OpData01.cpp]: It's usually called plaintext when it's input to a cipher or output of a decipher.

File: src/format/OpData01.cpp [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+        }
+        clearText = new QByteArray(actualClearText, len);
+        // qDebug("processInPlace -> <<%s>>\n", clearText->toHex(' ').data());
+    }
+    auto result = new OpData01();
```
Comment (phoerious) [inline src/format/OpData01.cpp]: Again, this smells like memory leaks.

File: src/format/OpData01.cpp [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+    result->m_clearText = clearText;
+    return result;
+}
+
+QByteArray* OpData01::getClearText()
+{
+    return this->m_clearText;
```
Comment (phoerious) [inline src/format/OpData01.cpp]: I'd name this `m_plaintext` and `getPlaintext()`.

File: src/format/OpData01.cpp [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+
+QString* OpData01::errorString()
+{
+    return this->m_errorStr;
+}
+
+OpData01::~OpData01()
+{
+    delete m_clearText;
+    delete m_errorStr;
```
Comment (phoerious) [inline src/format/OpData01.cpp]: This would be unnecessary if you used `QScopedPointer`.

File: src/format/OpVaultReader.cpp [hunk 19]
```diff
@@ -0,0 +1,498 @@
...
+#include <gcrypt.h>
+
+static void populateCategoryGroups(Group* rootGroup);
+
+OpVaultReader::~OpVaultReader()
+{
+    delete m_masterKey;
+    delete m_masterHmacKey;
+    delete m_overviewKey;
+    delete m_overviewHmacKey;
```
Comment (phoerious) [inline src/format/OpVaultReader.cpp]: Use `QScopedPointer` for all these.

File: src/format/OpVaultReader.cpp [hunk 20]
```diff
@@ -0,0 +1,498 @@
...
+    auto result = new Database();
+    readDatabase(opdataDir, result, passPhrase, passPhraseLen);
+    return result;
+}
+
+void OpVaultReader::readDatabase(QDir& opVaultDir, Database* db, const void* passPhrase, size_t passPhraseLen)
+{
+    if (!opVaultDir.exists()) {
+        m_error = true;
+        m_errorStr = "Directory .opvault must exist";
```
Comment (phoerious) [inline src/format/OpVaultReader.cpp]: I don't think the class should keep state for error messages. They should parameters passed by reference. Also, all messages need to be translatable.

File: src/format/OpVaultReader.cpp [hunk 21]
```diff
@@ -0,0 +1,498 @@
...
+    }
+
+    m_db = db;
+    m_meta = new Metadata(m_db);
+    auto rootGroup = new Group();
+    {
+        TimeInfo ti;
+        rootGroup->setTimeInfo(ti);
+        rootGroup->setUpdateTimeinfo(false);
+        rootGroup->setName("OPVault Root Group");
```
Comment (phoerious) [inline src/format/OpVaultReader.cpp]: I don't think that it should deviate from the standard KeePassXC root group name, but in any other case, it'd need to be translatable.

File: src/format/OpVaultReader.cpp [hunk 22]
```diff
@@ -0,0 +1,498 @@
...
+        m_errorStr = "Directory .opvault/default must be readable";
+        return;
+    }
+
+    m_db = db;
+    m_meta = new Metadata(m_db);
+    auto rootGroup = new Group();
+    {
+        TimeInfo ti;
+        rootGroup->setTimeInfo(ti);
```
Comment (phoerious) [inline src/format/OpVaultReader.cpp]: No need for a scope here. Just instantiate `TimeInfo` in-place.

File: src/format/OpVaultReader.cpp [hunk 23]
```diff
@@ -0,0 +1,498 @@
...
+        rootGroup->setUpdateTimeinfo(false);
+        rootGroup->setName("OPVault Root Group");
+        rootGroup->setUuid(QUuid::createUuid());
+    }
+    this->m_db->setRootGroup(rootGroup);
```
Comment (phoerious) [inline src/format/OpVaultReader.cpp]: Remove explicit `this`. I'd also prefer moving this directly below the allocation, so it's clear where this free-store object is parented.

File: src/format/OpVaultReader.cpp [hunk 24]
```diff
@@ -0,0 +1,498 @@
...
+        const QString bandChars("0123456789ABCDEF");
+        QString bandPattern("band_%1.js");
+        for (QChar ch : bandChars) {
+            QFile bandFile(defaultDir.filePath(bandPattern.arg(ch)));
+            if (!bandFile.exists()) {
+                qWarning() << "Skipping missing file \"" << bandFile.fileName() << "\"\n";
+                continue;
+            }
+            QJsonObject bandJs = readAndAssertJsonFile(bandFile, "ld(", ");");
+            qDebug() << "band " << ch << " contains " << bandJs.size() << " keys\n";
```
Comment (phoerious) [inline src/format/OpVaultReader.cpp]: Remove `qDebug()`.

File: src/format/OpVaultReader.cpp [hunk 25]
```diff
@@ -0,0 +1,498 @@
...
+
+#include "core/Group.h"
+#include "core/Tools.h"
+#include "crypto/CryptoHash.h"
+#include "crypto/SymmetricCipher.h"
+
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QUuid>
+#include <QtGui/QtGui>
```
Comment (phoerious) [inline src/format/OpVaultReader.cpp]: Use `<QtGlobal>` instead.

File: src/format/OpVaultReader.cpp [hunk 26]
```diff
@@ -0,0 +1,498 @@
...
+                const QString uuid = bandEnt["uuid"].toString();
+                if (entryKey != uuid) {
+                    qWarning() << QString("Mismatched Entry UUID, its JSON key <<%1>> and its UUID <<%2>>")
+                                      .arg(entryKey)
+                                      .arg(uuid);
+                }
+                QStringList requiredKeys;
+                requiredKeys << "d"
+                             << "k"
+                             << "hmac";
```
Comment (phoerious) [inline src/format/OpVaultReader.cpp]: This fits on one line.

File: src/format/OpVaultReader.cpp [hunk 27]
```diff
@@ -0,0 +1,498 @@
...
+                bool ok = true;
+                for (QString& requiredKey : requiredKeys) {
```
Comment (phoerious) [inline src/format/OpVaultReader.cpp]: `asConst(requiredKeys)`.

File: src/format/OpVaultReader.cpp [hunk 28]
```diff
@@ -0,0 +1,498 @@
...
+}
+
+QString OpVaultReader::errorString()
+{
+    return m_errorStr;
+}
+
+bool OpVaultReader::processProfileJson(QJsonObject& profileJson, const void* passPhrase, size_t passPhraseLen)
+{
+    Group* rootGroup = this->m_db->rootGroup();
```
Comment (phoerious) [inline src/format/OpVaultReader.cpp]: No explicit `this`.

**Comment by mdaniel** [comment]
I am working on incorporating these changes, and I am grateful for all the feedback. I just didn't want you to think the feedback went to `/dev/null`. I'll of course fix those conflicts, too.

**Comment by droidmonkey** [comment]
@mdaniel any update for this? Would love to see this in 2.4.0, but needs to be ready within the next couple weeks.

**Comment by mdaniel** [comment]
My apologies, I let this slip off my radar; I'll try to get those concerns cleaned up, and fix the merge conflicts.

**Comment by droidmonkey** [comment]
@mdaniel still working on this?

**Comment by mdaniel** [comment]
Wow, I had no idea it had been 20 days since that comment :-( But I am hoping Friday will be a solid block of time where I can focus, as that's been the obstacle

File: src/format/OpData01.cpp (hunk 1 above)
Comment (mdaniel) [reply src/format/OpData01.cpp]: > I believe the minimum supported version (per INSTALL.md) is Qt 5.2

Related to that, the state of `develop` (and `2.3.4`) against 5.2 is in rough shape.

[`QAtomicInteger`](http://doc.qt.io/qt-5.6/qatomicinteger.html) was introduced in 5.3 but is used by [`src/browser/NativeMessagingBase.h`](https://github.com/keepassxreboot/keepassxc/blob/2.3.4/src/browser/NativeMessagingBase.h#L31). Replacing it with `QAtomicInt` makes the file at least compile, but then it surfaces this next:

```
/opt/qt52/include/QtCore/qjsonvalue.h: In member function 'QJsonObject BrowserAction::handleSetLogin(const QJsonObject&, const QString&)':
/opt/qt52/include/QtCore/qjsonvalue.h:119:12: error: 'QJsonValue::QJsonValue(const void*)' is private
     inline QJsonValue(const void *) {}
            ^
/keepassxc/src/browser/BrowserAction.cpp:332:22: error: within this context
     message["error"] = "";
```
Wrapping those literals in `QString()` gets us past that, but then:

```
/keepassxc/src/browser/BrowserService.cpp: In member function 'QJsonArray BrowserService::findMatchingEntries(const QString&, const QString&, const QString&, const QString&, const StringPairList&)':
/keepassxc/src/browser/BrowserService.cpp:265:16: error: no match for 'operator<<' (operand types are 'QJsonArray' and 'QJsonObject')
         result << prepareEntry(entry);
                ^
/keepassxc/src/browser/BrowserService.cpp:265:16: note: candidates are:
In file included from /opt/qt52/include/QtCore/qstring.h:45:0,
                 from /opt/qt52/include/QtCore/qjsonvalue.h:46,
                 from /opt/qt52/include/QtCore/qjsonarray.h:45,
                 from /opt/qt52/include/QtCore/QJsonArray:1,
                 from /keepassxc/src/browser/BrowserService.cpp:20:
/opt/qt52/include/QtCore/qchar.h:521:28: note: QDataStream& operator<<(QDataStream&, QChar)
```
changing to `.append()` cures that

---

One can see this for themselves via:
```diff
diff --git a/Dockerfile b/Dockerfile
index e1fd26d4..89a30289 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -18,8 +18,8 @@ FROM ubuntu:14.04
 
 ENV REBUILD_COUNTER=10
 
-ENV QT5_VERSION=qt510
-ENV QT5_PPA_VERSION=qt-5.10.1
+ENV QT5_VERSION=qt52
+ENV QT5_PPA_VERSION=qt521
 ENV TERM=xterm-256color
 
 RUN set -x \
@@ -55,6 +55,7 @@ RUN set -x \
         libxi-dev \
         libxtst-dev \
         mesa-common-dev \
+        libgl1-mesa-dev \
         libyubikey-dev \
         libykpers-1-dev \
         libqrencode-dev \
```
(which kind of surprises me isn't the default build configuration, given the stated minimum dep)

I can create a PR against `develop` to cure those, if you'd like, but I wanted to make you aware.

**Comment by mdaniel** [comment]
I took a light swing at replacing `Foo*` with `QSharedPointer<Foo>` but there is no way I'll get that to even run tonight. So, I guess we can put this back in the icebox until I have the mental bandwidth to learn `QSharedPointer` or have that as a stretch goal

**Comment by mdaniel** [comment]
> In addition to these comments, I have some general remarks I'd like you to think about:
First, this is a lot of code only for reading a password file. Our KDBX code is a fraction of that. Is it really necessary to merge so much code only for adding another file format?

Up to you (collectively). I only put this in flight because I thought others would benefit from someone taking the time to actually implement the OpVault reader. I actually [did mention](https://github.com/keepassxreboot/keepassxc/issues/1462#issuecomment-364019431) that this kind of thing is perfect for a plug-in, but it was my understanding that KeePassXC has no such concept, so that's how I ended up wiring it into this repo.

> Secondly, please don't use so many sub scopes. Better think about the variables you need and use rvalues where applicable. Too many indentations make the code harder to grasp and lead to sloppy variable use.

I don't know what an rvalues is, and I readily and wholeheartedly admit that the structure of the code is a lot more chaotic than I'd like. I can't say offhand how much bandwidth I'll have to straighten that stuff out, although it's a lot more likely in deep December early January.

**Comment by mdaniel** [comment]
Finally, I wanted to point out that

> My code follows the code style of this project. (FYI, make format dirties 103 files that I didn't touch)

is still true and makes keeping my changes in sync with `clang-format` a **monster** PITA since the workflow is more or less `make format && git checkout $(all files that are not mine)`

**Comment by droidmonkey** [comment]
I was wondering why you were having so many format issues and I was reminded that we didn't merge the code cleanup part 2 PR. That is now merged!

**Comment by user13542** [comment]
@mdaniel any news?

**Comment by Jorijn** [comment]
I'm also eagerly waiting for this feature to be completed.

**Comment by mdaniel** [comment]
I finally realized why this PR currently feels overwhelming: I don't have a list of _tactical_ changes that I would make to it. I am not a C++ developer, so I don't have a deep understanding of `QSharedPointer<>` or its hazards, and as I mentioned, I don't know what an rvalue is.

If phoerious (et al) could lay out the criteria for accepting this, then it might make it a little more approachable, since -- aside from the sub-scopes and generally monolithic functions -- I don't know what I would open my editor and do in order to clear the merge blocker

**Comment by droidmonkey** [comment]
Frankly this needs to be built from the "ground up" again. With the database refactor and methods you used it might not be easily salvageable in this state.

**Comment by mdaniel** [comment]
FWIW, I just merged in develop (0f1be60b) into my branch and it built cleanly and behaved as I expected (it even looks nicer; maybe the icons got a refresh?), so I don't know if the database refactor you mention is still pending, or is on develop already

I guess once I learn when the refactor is coming, or more about that, then I can evaluate how to proceed. But if you're unhappy with this branch, then I guess we'll just close this so its state will be more obvious to those waiting for it to land.

**Comment by droidmonkey** [comment]
Ah good to know, I guess you started this after the refactor was merged. I take back my statement.

**Comment by droidmonkey** [comment]
I am taking this under my wing. I have squashed all the extraneous commits and split the non-opvault code changes from opvault. I took a first round at getting rid of all the raw pointers and primitive objects.

This still needs A TON of work, but it does open an OPVault at this time. It does not save the resulting database and there are crashes.

**Comment by mdaniel** [comment]
> It does not save the resulting database

That wasn't ever one of my objectives, as I wasn't trying to _import_ the opvault, just open it for use

I let this linger because I wasn't sure what action to take on it to make it move forward, so I'm glad you have something specific in mind. Let me know how I can help, even if it's just sending you good vibes :-)

**Comment by droidmonkey** [comment]
OK the code is pretty much where I want it to be right now. I also fixed saving the database.

The only other thing I want to do is actually move some of the attributes into username/password/notes instead of cramming everything into incredibly unhelpful attributes.

**Comment by droidmonkey** [comment]
OK I believe this to be ready for prime time. Going to let CI work its magic then merge when its over. Anyone with 1Password, I would really appreciate you testing the import feature once the snapshot is available.

**Comment by mdaniel** [comment]
I **sincerely** appreciate your help in shepherding this change in, and I'm glad it didn't take a monster amount of rewrite to bring it up to your satisfaction

**Comment by droidmonkey** [comment]
You made a very good core! Thank you for bringing this in, a lot of people will benefit from this feature.

**Comment by Jorijn** [comment]
I just tested this feature on a mac. Seems to work as intended. Kudos for even including the MFA tokens. 

Just a small remark: On Macs, .opvault directories are registered as "files". I was unable to select it until I removed the ".opvault" part to turn it into a directory again.

**Comment by droidmonkey** [comment]
That's interesting, we could display a file selection dialog on macOS instead of a directory selection.

**Comment by Zettt** [comment]
Guys, great work first of all, but I'm wondering why you're messing with the opvault, which normal users cannot easily access? Why is it not better to simply support the normal (1Password preferred) way of exchanging this information through 1pif files?

**Comment by droidmonkey** [comment]
Great "now you tell us" lol. Is 1pif wven documented?

**Comment by Zettt** [comment]
You guys didn't know this? Oh shishkebab. To be honest I have no idea how it is documented at all. All I can say is that from what I've read here I wouldn't even be able to test the feature. I'm on their subscription plan. As far as I'm aware, there's no real way to access the vault at all. *But*, and that's the reason I've commented, there's an entry in the file menu which says Export → All Items. The 1pif format is in JSON-ish format. You guys might have luck to check out the Bitwarden repositories, they also have an importer for the format. I'm not a programmer by day, but I work as manager for teams. My gut feeling tells that what the 1pif has in it, is probably just the already-encrypted content of the opvault. Perhaps. Or so.

Here's an example:

```JSON
{"uuid":"vthnyeitfvb6pc6iyc4a5i7rrm","updatedAt":1559854808,"locationKey":"riot.app","securityLevel":"SL5","contentsHash":"92c59ef6","title":"Riot","location":"app:\/\/im.riot.app","secureContents":{"URLs":[{"url":"app:\/\/im.riot.app"},{"label":"","url":"riot.im"}],"fields":[{"value":"obviously this is where the password goes","name":"password","type":"P","designation":"password"},{"value":"azeitler","name":"username","type":"T","designation":"username"}],"sections":[{"title":"Related Items","name":"linked items"}]},"createdAt":1559853537,"typeName":"webforms.WebForm"}
***5642bee8-a5ff-11dc-8314-0800200c9a66***
```

**Comment by droidmonkey** [comment]
Let's continue this discussion over in #3239

**Pull request merged**